File: Hackerrank/sherlockAndCost.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the cost function below.
def cost(b):
    n=len(b)
    l, h = 0, 0
    for i in range(1, n):
        l, h = (max(l, h + b[i - 1] - 1),
                max(l + b[i] - 1, h + abs(b[i] - b[i - 1])))
    return max(l, h)

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    t = int(input())

    for t_itr in range(t):
        n = int(input())

        B = list(map(int, input().rstrip().split()))

        result = cost(B)

        fptr.write(str(result) + '\n')

    fptr.close()

File: Hackerrank/sherlock-and-cost.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> split_string(string);

int cost(vector<int> b) {
  int low = 0;
  int high = 0;
  int maxi = 0;
  for (int i = 1; i < b.size(); i++) {
    int prev_low = low;
    int prev_high = high;
    low = max(prev_low, high + abs(b[i - 1] - 1));
    high = max(prev_high + abs(b[i] - b[i - 1]), prev_low + abs(b[i] - 1)); 
  }
  return max(low, high);
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int t;
    cin >> t;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    for (int t_itr = 0; t_itr < t; t_itr++) {
        int n;
        cin >> n;
        cin.ignore(numeric_limits<streamsize>::max(), '\n');

        string B_temp_temp;
        getline(cin, B_temp_temp);

        vector<string> B_temp = split_string(B_temp_temp);

        vector<int> B(n);

        for (int i = 0; i < n; i++) {
            int B_item = stoi(B_temp[i]);

            B[i] = B_item;
        }

        int result = cost(B);

        fout << result << "\n";
    }

    fout.close();

    return 0;
}

vector<string> split_string(string input_string) {
    string::iterator new_end = unique(input_string.begin(), input_string.end(), [] (const char &x, const char &y) {
        return x == y and x == ' ';
    });

    input_string.erase(new_end, input_string.end());

    while (input_string[input_string.length() - 1] == ' ') {
        input_string.pop_back();
    }

    vector<string> splits;
    char delimiter = ' ';

    size_t i = 0;
    size_t pos = input_string.find(delimiter);

    while (pos != string::npos) {
        splits.push_back(input_string.substr(i, pos - i));

        i = pos + 1;
        pos = input_string.find(delimiter, i);
    }

    splits.push_back(input_string.substr(i, min(pos, input_string.length()) - i + 1));

    return splits;
}
